sdk/dotnet/Image.cs

```diff
@@ -0,0 +1,170 @@
+        /// </summary>
+        [Output("imageName")]
+        public Output<string> ImageName { get; private set; }
+
+        /// <summary>
+        /// The server the image is located at.
+        /// </summary>
+        [Output("registryServer")]
+        public Output<string?> RegistryServer { get; private set; }
+
+        public Image(string name, ImageArgs args, ComponentResourceOptions? options = null)
+            : base("docker:image:Image", name, options)
+        {
+            this.ImageName = Output.Tuple(args.ImageName, args.LocalImageName.Unwrap()).Apply(imageArgs =>
+            {
+                var imageName = imageArgs.Item1;
+
+                // If there is no localImageName set it equal to imageName.  Note: this means
+                // that if imageName contains a tag, localImageName will contain the same tag.
+                var localImageName = imageArgs.Item2 ?? imageName;
+
+                // Now break both the localImageName and the imageName into the untagged part and the
+                // optional tag.  If both have tags, they must match.  If one or the other has a tag, we
+                // just use that as the tag to use.  This allows users to flexibly provide a tag on one
+                // option or the other and still have it work out.
+                var (localImageNameWithoutTag, localImageNameTag) = Utils.GetImageNameAndTag(localImageName);
+                var (imageNameWithoutTag, imageNameTag) = Utils.GetImageNameAndTag(imageName);
+
+                if (localImageNameTag != null && imageNameTag != null && localImageNameTag != imageNameTag)
+                    throw new InvalidOperationException(
+                        $"[{nameof(args.LocalImageName)}] and [{nameof(args.ImageName)}] had mismatched tags. {localImageNameTag} != {imageNameTag}");
+
+                var tag = localImageNameTag ?? imageNameTag;
+
+                // BuildAndPushImageAsync expects only the baseImageName to have a tag.  So build that
+                // name appropriately if we were given a tag.
+                var baseImageName = tag != null ? $"{localImageNameWithoutTag}:{tag}" : localImageName;
+
+                // BuildAndPushImageAsync does not want the repositoryUrl to have a tag.  This is just
+                // the base url where the images will be pushed to.  All tagging will be taken care of
+                // inside that api.
+                var repositoryUrl = imageNameWithoutTag;
+                return Docker.BuildAndPushImageAsync(
+                    baseImageName,
+                    args.Build,
+                    repositoryUrl,
+                    this,
+                    args.Registry);
+            });
+
+            this.RegistryServer = args.Registry != null ? args.Registry.Apply(r => r.Server).Apply(v => (string?)v) : Output.Create((string?)null);
+            this.BaseImageName = args.ImageName;
+
+            this.RegisterOutputs(
```

Comment:
note: general philosophy now is that components don't actually regiter any outputs.. We just use .REgisterOutputs to indicate when we're done.

Reply:
I started with that but it wouldn't work: the `ImageName = ...Apply` would never be called

Reply:
ah interesting.  i solve that *inside* Pulumi through the use of `IRunner.RegistrTask`.  However, i don't really want to expose that to people. 

An alternate model is that every `.Apply` should register the task to ensure that we never close the program while there are outstanding tasks (that we own) in flight.

Reply:
For now, your approach is a good way to handle things!